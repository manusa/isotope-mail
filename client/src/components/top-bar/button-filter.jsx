import React from 'react';
import {connect} from 'react-redux';
import {translate} from 'react-i18next';
import TopBarButton from './top-bar-button';
import FilterDialog from './filter-dialog';
import {messageFilterActive} from '../../selectors/application';
import mainCss from '../../styles/main.scss';
import styles from './button-filter.scss';

export class ButtonFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      dialogVisible: false
    };
    this.handleOnToggleDialog = this.onToggleDialog.bind(this);
    this.handleOnCloseDialog = this.onCloseDialog.bind(this);
  }

  render() {
    const {t, active} = this.props;
    const {dialogVisible} = this.state;
    return <span
      className={`${styles['button-filter']} ${mainCss['mdc-menu-surface--anchor']}`}
      isotip={t('topBar.quickFilter')} isotip-position='bottom-end' isotip-size='small'
      isotip-hidden={dialogVisible.toString()}>
      <TopBarButton
        className={`${styles['button-filter--button']} ${active ? styles.active : ''}`}
        onClick={this.handleOnToggleDialog}>filter_list</TopBarButton>
      <FilterDialog visible={dialogVisible} closeFilterDialogHandler={this.handleOnCloseDialog} />
    </span>;
  }

  componentDidMount() {
    window.addEventListener('click', this.handleOnCloseDialog);
  }

  componentWillUnmount() {
    window.removeEventListener('click', this.handleOnCloseDialog);
  }

  onToggleDialog(event) {
    this.setState(previousState => ({dialogVisible: !previousState.dialogVisible}));
    event.stopPropagation();
  }

  onCloseDialog() {
    if (this.handleOnCloseDialog.disabled !== true) {
      this.setState({dialogVisible: false});
    }
    delete this.handleOnCloseDialog.disabled;
  }
}

const mapStateToProps = state => ({
  active: messageFilterActive(state)
});

export default connect(mapStateToProps)(translate()(ButtonFilter));
