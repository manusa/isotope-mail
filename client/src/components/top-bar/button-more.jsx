import React, {useState, useEffect} from 'react';
import {translate} from 'react-i18next';
import TopBarButton from './top-bar-button';
import mainCss from '../../styles/main.scss';

export const ButtonMore = ({t, children}) => {
  const [menuOpen, setMenuOpen] = useState(false);
  useEffect(() => {
    function closeMenu() {
      setMenuOpen(false);
    }
    window.addEventListener('click', closeMenu);
    return function unmount() {
      window.removeEventListener('click', closeMenu);
    };
  }, []);
  return (
    <span
      className={`${mainCss['mdc-menu-surface--anchor']}`}
      isotip={t('topBar.more')} isotip-position='bottom-end' isotip-size='small'
      isotip-hidden={menuOpen.toString()}>
      <TopBarButton
        onClick={event => {
          setMenuOpen(!menuOpen);
          event.stopPropagation();
        }}>more_vert</TopBarButton>
      {React.cloneElement(children, {visible: menuOpen})}
    </span>
  );
};

export default translate()(ButtonMore);
