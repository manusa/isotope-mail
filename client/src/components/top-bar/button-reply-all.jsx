import React from 'react';
import PropTypes from 'prop-types';
import {translate} from 'react-i18next';
import TopBarButton from './top-bar-button';

export const ButtonReplyAll = ({t, outboxEmpty, replyAllMessage}) => (
  outboxEmpty && (
    <span isotip={t('topBar.replyAll')} isotip-position='bottom' isotip-size='small'>
      <TopBarButton onClick={replyAllMessage}>reply_all</TopBarButton>
    </span>)
);

ButtonReplyAll.propTypes = {
  outboxEmpty: PropTypes.bool.isRequired,
  replyAllMessage: PropTypes.func.isRequired
};

export default translate()(ButtonReplyAll);
