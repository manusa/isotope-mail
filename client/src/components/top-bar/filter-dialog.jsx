import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {translate} from 'react-i18next';
import {
  setMessageFilterKey as setMessageFilterKeyAction,
  setMessageFilterText as setMessageFilterTextAction
} from '../../actions/application';
import MessageFilters from '../../services/message-filters';
import styles from './filter-dialog.scss';
import mainCss from '../../styles/main.scss';
import {
  activeMessageFilter as activeMessageFilterSelector,
  messageFilterText as messageFilterTextSelector} from '../../selectors/application';

export const FilterDialog = (
  {
    t, visible, closeFilterDialogHandler,
    activeMessageFilter, setMessageFilterKey,
    messageFilterText, setMessageFilterText
  }) =>
  <div
    className={`${styles['filter-dialog']} ${mainCss['mdc-menu']} ${mainCss['mdc-menu-surface']}
    ${visible ? mainCss['mdc-menu-surface--open'] : ''}`}
    aria-hidden={!visible}
  >
    <ul className={`${mainCss['mdc-list']} ${mainCss['mdc-list--dense']}`} >
      <li
        className={`${styles['filter-dialog__item-text']} ${messageFilterText ?
          styles['filter-dialog__item-text--active'] : ''} ${mainCss['mdc-list-item']}`}
      >
        <input
          className={styles['text-field']}
          value={messageFilterText}
          onChange={e => setMessageFilterText(e.target.value)}
          onMouseDown={() => {
            closeFilterDialogHandler.disabled = true;
          }}
        />
        <i className={`${styles.icon} material-icons ${mainCss['mdc-button__icon']}`}>search</i>
      </li>
      {Object.entries(MessageFilters).map(([key, value]) => {
        const active = key === activeMessageFilter.key;
        return (
          <li
            key={key}
            className={`${styles['filter-dialog__item-check']} ${mainCss['mdc-list-item']}
          ${active ? mainCss['mdc-list-item--selected'] : ''}`}
            onClick={() => setMessageFilterKey(value)}
          >
            <i className={`${styles.check} ${active ? styles['check--active'] : ''} material-icons`}>
              check
            </i>
            {t(value.i18nKey)}
          </li>
        );
      })}
    </ul>
  </div>;

FilterDialog.propTypes = {
  t: PropTypes.func,
  visible: PropTypes.bool.isRequired,
  closeFilterDialogHandler: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  activeMessageFilter: activeMessageFilterSelector(state),
  messageFilterText: messageFilterTextSelector(state)
});

const mapDispatchToProps = dispatch => ({
  setMessageFilterKey: messageFilter => dispatch(setMessageFilterKeyAction(messageFilter.key)),
  setMessageFilterText: messageFilterText => dispatch(setMessageFilterTextAction(messageFilterText))
});

export default connect(mapStateToProps, mapDispatchToProps)(translate()(FilterDialog));
