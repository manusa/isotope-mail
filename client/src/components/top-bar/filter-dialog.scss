@import '../../styles/partials/variables';
@import '@material/list/variables';

$icon-size: 24px;

.filter-dialog {
  right: 0px;
  min-width: 150px;
  z-index: map-get($z-index, thermosphere);

  &__item-text {
    position: relative;
    padding: 0;
    z-index: 0;
    .text-field {
      z-index: 2;
      outline: 0;
      border: 0;
      padding: 0 $mdc-list-side-padding;
      height: 100%;
      line-height: 100%;
      background: transparent;
    }
    .icon {
      position: absolute;
      z-index: 1;
      font-size: $icon-size;
      left: $mdc-list-side-padding;
      transition: left 200ms ease-out;
    }
    .text-field:focus + .icon,
    &--active .icon {
      left: calc(100% - #{$icon-size} - #{$mdc-list-side-padding});
    }
  }
  &__item-check{
    .check {
      font-size: 18px;
      padding-right: 1rem;
      opacity: 0.2;

      &--active {
        color: $mdc-theme-primary;
        opacity: 1;
      }
    }
  }
}
