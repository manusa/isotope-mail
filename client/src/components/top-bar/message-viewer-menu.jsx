import React, {useState} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {translate} from 'react-i18next';
import {getSelectedFolder} from '../../selectors/folders';
import {getCredentials} from '../../selectors/application';
import {downloadMessage as downloadMessageService} from '../../services/message';
import mainCss from '../../styles/main.scss';

export const DownloadListItem = ({t, downloadMessage}) => {
  const [downloading, setDownloading] = useState(false);
  const onDownloadClick = async () => {
    if (!downloading) {
      setDownloading(true);
      await downloadMessage();
      setDownloading(false);
    }
  };
  return (
    <li
      className={`${mainCss['mdc-list-item']} ${downloading ? mainCss['mdc-list-item--disabled'] : ''}`}
      onClick={onDownloadClick}
    >{t('topBar.messageViewerMenu.download')}</li>
  );
};

export const ListUnsubscribeListItem = ({t, message}) => {
  let listItem = null;
  if (message.listUnsubscribe && message.listUnsubscribe.length > 0) {
    const entries = message.listUnsubscribe[0].match(/<([^>]*?)>/);
    if (entries) {
      listItem = (
        <li>
          <a
            className={mainCss['mdc-list-item']}
            href={entries[1]} target="_blank"
          >
            {t('topBar.messageViewerMenu.listUnsubscribe')}
          </a>
        </li>
      );
    }
  }
  return listItem;
};

export const MessageViewerMenu = ({t, visible, selectedFolder, selectedMessage, downloadMessage}) =>
  selectedFolder && selectedMessage && (
    <div
      className={`${mainCss['message-viewer-menu']} ${mainCss['mdc-menu']} ${mainCss['mdc-menu-surface']}
      ${visible ? mainCss['mdc-menu-surface--open'] : ''}`}
      aria-hidden={!visible}
    >
      <ul className={`${mainCss['mdc-list']} ${mainCss['mdc-list--dense']}`} >
        <DownloadListItem t={t} downloadMessage={downloadMessage}/>
        <ListUnsubscribeListItem t={t} message={selectedMessage}/>
      </ul>
    </div>
  );

MessageViewerMenu.propTypes = {
  t: PropTypes.func,
  visible: PropTypes.bool,
  credentials: PropTypes.object,
  selectedFolder: PropTypes.object,
  selectedMessage: PropTypes.object,
  downloadMessage: PropTypes.func
};

const mapStateToProps = state => ({
  credentials: getCredentials(state),
  selectedFolder: getSelectedFolder(state) || null,
  selectedMessage: state.application.selectedMessage
});

const mergeProps = (stateProps, dispatchProps, ownProps) => (Object.assign({}, stateProps, dispatchProps, ownProps, {
  downloadMessage: async () =>
    downloadMessageService(stateProps.credentials, stateProps.selectedFolder, stateProps.selectedMessage)
}));

export default connect(mapStateToProps, null, mergeProps)(translate()(MessageViewerMenu));
