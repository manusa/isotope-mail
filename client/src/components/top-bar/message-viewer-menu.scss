@import '../../styles/partials/variables';

.message-viewer-menu {
  right: 0;
  min-width: 150px;
  z-index: map-get($z-index, thermosphere);
}
