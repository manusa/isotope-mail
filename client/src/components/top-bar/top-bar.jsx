import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import ConfirmDeleteFromTrashDialog from './confirm-delete-from-trash-dialog';
import TopBarMessageList from './top-bar-message-list';
import TopBarMessageViewer from './top-bar-message-viewer';
import TopBarMessageEditor from './top-bar-message-editor';
import {getCredentials, selectedMessage as selectedMessageSelector} from '../../selectors/application';
import {getSelectedFolder} from '../../selectors/folders';
import {selectedFolderMessagesFilteredAndSelected, selectedMessagesIds} from '../../selectors/messages';
import {findTrashFolder, FolderTypes} from '../../services/folder';
import {forwardMessage, replyMessage, clearSelectedMessage} from '../../services/application';
import {deleteMessages, moveMessages, setMessagesSeen} from '../../services/message';
import mainCss from '../../styles/main.scss';

export class TopBar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      deletingFromTrash: false,
      deletingFromTrashConfirm: () => {}
    };
  }

  render() {
    const props = this.props;
    const {sideBarToggle, selectedMessages, selectedMessagesAllUnread, outbox, toggleMessageSeen} = props;
    const collapsed = props.sideBarCollapsed;
    const isEditing = props.newMessage && Object.keys(props.newMessage).length > 0;
    const isMessageViewer = props.selectedMessage && Object.keys(props.selectedMessage).length > 0;
    let title = props.title;
    if (props.selectedFolder && props.selectedFolder.name
      && props.selectedFolder.type !== FolderTypes.INBOX) {
      title = `${props.selectedFolder.name} - ${title}`;
    }
    return (
      <header className={`
      ${collapsed ? '' : mainCss['mdc-top-app-bar--with-side-bar']}
      ${mainCss['mdc-top-app-bar']} ${mainCss['mdc-top-app-bar--fixed']}`}>
        {!isEditing && !isMessageViewer
          && (<TopBarMessageList
            title={title} collapsed={collapsed} sideBarToggle={sideBarToggle}
            selectedMessages={selectedMessages}
            onDeleteClick={() => this.onDelete(props.deleteMessages)}
            selectedMessagesAllUnread={selectedMessagesAllUnread}
            onMarkReadClick={() => props.setMessagesSeen(true)}
            onMarkUnreadClick={() => props.setMessagesSeen(false)}
          />)
        }
        {!isEditing && isMessageViewer
          && (<TopBarMessageViewer
            collapsed={collapsed} sideBarToggle={sideBarToggle} clearSelectedMessage={props.clearSelectedMessage}
            outboxEmpty={outbox === null}
            onReplyMessageClick={props.replyMessage} onForwardMessageClick={props.forwardMessage}
            onDeleteClick={() => this.onDelete(props.deleteMessage)} onMarkUnreadClick={toggleMessageSeen}/>)
        }
        {isEditing
          && (<TopBarMessageEditor
            title={title} collapsed={collapsed} sideBarToggle={sideBarToggle}
          />)
        }
        <ConfirmDeleteFromTrashDialog
          visible={this.state.deletingFromTrash}
          deleteAction={this.state.deletingFromTrashConfirm}
          cancelAction={() => this.setState({deletingFromTrash: false})}
        />
      </header>
    );
  }

  onDelete(action) {
    if (this.props.selectedFolder.type === FolderTypes.TRASH) {
      this.setState({
        deletingFromTrash: true,
        deletingFromTrashConfirm: () => {
          action();
          this.setState({deletingFromTrash: false});
        }
      });
    } else {
      action();
    }
  }
}

TopBar.propTypes = {
  title: PropTypes.string,
  newMessage: PropTypes.object,
  selectedFolder: PropTypes.object,
  selectedMessagesIds: PropTypes.array,
  selectedMessages: PropTypes.array,
  selectedMessage: PropTypes.object,
  selectedMessagesAllUnread: PropTypes.bool,
  clearSelectedMessage: PropTypes.func,
  sideBarToggle: PropTypes.func.isRequired,
  sideBarCollapsed: PropTypes.bool.isRequired
};

const mapStateToProps = state => {
  const selectedMessages = selectedFolderMessagesFilteredAndSelected(state);
  const selectedMessagesAllUnread = selectedMessages.filter(m => m.seen === true).length === 0;
  return ({
    title: state.application.title,
    newMessage: state.application.newMessage,
    outbox: state.application.outbox,
    selectedFolder: getSelectedFolder(state) || null,
    selectedMessagesIds: selectedMessagesIds(state),
    selectedMessages: selectedMessages,
    selectedMessage: selectedMessageSelector(state),
    selectedMessagesAllUnread: selectedMessagesAllUnread,
    credentials: getCredentials(state),
    folders: state.folders
  });
};

const mapDispatchToProps = dispatch => ({
  clearSelectedMessage: () => clearSelectedMessage(dispatch),
  replyMessage: selectedMessaage => replyMessage(dispatch, selectedMessaage),
  forwardMessage: selectedMessaage => forwardMessage(dispatch, selectedMessaage),
  deleteMessage: (credentials, folders, selectedFolder, selectedMessage) => {
    const trashFolder = findTrashFolder(folders);
    if (selectedMessage && selectedFolder && trashFolder) {
      if (selectedFolder === trashFolder) {
        deleteMessages(dispatch, credentials, selectedFolder, [selectedMessage]);
      } else {
        moveMessages(dispatch, credentials, selectedFolder, trashFolder, [selectedMessage]);
      }
      clearSelectedMessage(dispatch);
    }
  },
  toggleMessageSeen: (credentials, selectedFolder, selectedMessage) => {
    setMessagesSeen(dispatch, credentials, selectedFolder, [selectedMessage], !selectedMessage.seen);
    clearSelectedMessage(dispatch);
  },
  deleteMessages: (credentials, folders, selectedFolder, selectedMessages) => {
    const trashFolder = findTrashFolder(folders);
    if (selectedMessages.length > 0 && selectedFolder && trashFolder) {
      if (selectedFolder === trashFolder) {
        deleteMessages(dispatch, credentials, selectedFolder, selectedMessages);
      } else {
        moveMessages(dispatch, credentials, selectedFolder, trashFolder, selectedMessages);
      }
    }
  },
  setMessagesSeen: (credentials, selectedFolder, selectedMessages, seen) => {
    if (selectedMessages.length > 0 && selectedFolder) {
      setMessagesSeen(dispatch, credentials, selectedFolder, selectedMessages, seen);
    }
  }
});

const mergeProps = (stateProps, dispatchProps, ownProps) => (Object.assign({}, stateProps, dispatchProps, ownProps, {
  replyMessage: () => dispatchProps.replyMessage(stateProps.selectedMessage),
  forwardMessage: () => dispatchProps.forwardMessage(stateProps.selectedMessage),
  deleteMessage: () =>
    dispatchProps.deleteMessage(
      stateProps.credentials, stateProps.folders, stateProps.selectedFolder, stateProps.selectedMessage),
  toggleMessageSeen: () =>
    dispatchProps.toggleMessageSeen(
      stateProps.credentials, stateProps.selectedFolder, stateProps.selectedMessage),
  deleteMessages: () =>
    dispatchProps.deleteMessages(
      stateProps.credentials, stateProps.folders, stateProps.selectedFolder, stateProps.selectedMessages),
  setMessagesSeen: seen =>
    dispatchProps.setMessagesSeen(
      stateProps.credentials, stateProps.selectedFolder, stateProps.selectedMessages, seen)
}));

export default connect(mapStateToProps, mapDispatchToProps, mergeProps)(TopBar);
