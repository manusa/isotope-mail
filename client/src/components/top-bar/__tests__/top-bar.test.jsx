import React from 'react';
import {shallow} from 'enzyme/build/index';
import TopBarConnected from '../top-bar';
import {TopBar} from '../top-bar';
import {FolderTypes} from '../../../services/folder';
import {createMockStore, MOCK_STORE} from '../../../__testutils__/store';
import * as applicationService from '../../../services/application';
import * as messgeService from '../../../services/message';
import {INITIAL_STATE} from '../../../reducers';

describe('TopBar component test suite', () => {
  describe('Snapshot render', () => {
    test('TopBar with selected folder and no new message and no selected message, should render in list mode', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedFolder: {type: FolderTypes.FOLDER, name: 'SelectedFolder'},
        selectedMessages: [],
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with selected folder (and messageIds) and no new message and no selected message, should render in list mode with delete and mark read buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedFolder: {type: FolderTypes.FOLDER, name: 'SelectedFolder'},
        selectedMessages: [{uid: 1}],
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with selected message and no new message and no selected folder, should render in view mode with buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        selectedMessage: {uid: 1},
        title: 'Isotope Mail Client - TITLE',
        outbox: null};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
    test('TopBar with new message and no selected message and no selected folder, should render in edit mode without buttons', () => {
      // Given
      const props = {
        sideBarCollapsed: false,
        sideBarToggle: jest.fn(),
        newMessage: {subject: 'New Message'},
        title: 'Isotope Mail Client - TITLE'};

      // When
      const topBar = shallow(<TopBar {...props}/>);

      // Then
      expect(topBar).toMatchSnapshot();
    });
  });
  describe('State properties tests', () => {
    test('clearSelectedMessage, application service function invoked', () => {
      // Given
      applicationService.clearSelectedMessage = jest.fn();
      const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
      const topBar = shallow(<TopBarConnected store={MOCK_STORE} {...props}/>);

      // When
      topBar.props().clearSelectedMessage();

      // Then
      expect(applicationService.clearSelectedMessage).toHaveBeenCalledTimes(1);
    });
    test('replyAllMessage, application service function invoked', () => {
      // Given
      const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application}};
      initialState.application.selectedMessage = {};
      const replyAllMessage = jest.fn(sm => expect(sm).toBe(initialState.application.selectedMessage));
      // noinspection JSUnresolvedVariable
      applicationService.replyAllMessage = jest.fn(() => replyAllMessage);
      const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
      const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

      // When
      topBar.props().replyAllMessage();

      // Then
      expect(applicationService.replyAllMessage).toHaveBeenCalledTimes(1);
      expect(replyAllMessage).toHaveBeenCalledTimes(1);
    });
    test('forwardMessage, application service function invoked', () => {
      // Given
      const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application}};
      initialState.application.selectedMessage = {};
      applicationService.forwardMessage = jest.fn((dispatch, sm) =>
        expect(sm).toBe(initialState.application.selectedMessage));
      const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
      const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

      // When
      topBar.props().forwardMessage();

      // Then
      expect(applicationService.forwardMessage).toHaveBeenCalledTimes(1);
    });
    describe('deleteMessage', () => {
      test('Selected message and selected folder (!== trash), messageService moveMessages function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE,
          application: {...INITIAL_STATE.application}, folders: {...INITIAL_STATE.folders}};
        initialState.application.selectedMessage = {};
        initialState.application.selectedFolderId = '1337';
        initialState.folders.explodedItems = {1337: {}, trash: {type: FolderTypes.TRASH}};
        applicationService.clearSelectedMessage = jest.fn();
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessage();

        // Then
        expect(messgeService.moveMessages).toHaveBeenCalledTimes(1);
        expect(messgeService.deleteMessages).not.toHaveBeenCalled();
        expect(applicationService.clearSelectedMessage).toHaveBeenCalledTimes(1);
      });
      test('Selected message and selected folder (== trash), messageService deleteMessages function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE,
          application: {...INITIAL_STATE.application}, folders: {...INITIAL_STATE.folders}};
        initialState.application.selectedMessage = {};
        initialState.application.selectedFolderId = '1337';
        initialState.folders.explodedItems = {1337: {type: FolderTypes.TRASH}};
        applicationService.clearSelectedMessage = jest.fn();
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessage();

        // Then
        expect(messgeService.deleteMessages).toHaveBeenCalledTimes(1);
        expect(messgeService.moveMessages).not.toHaveBeenCalled();
        expect(applicationService.clearSelectedMessage).toHaveBeenCalledTimes(1);
      });
      test('NO selected message, no function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE};
        applicationService.clearSelectedMessage = jest.fn();
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessage();

        // Then
        expect(messgeService.deleteMessages).not.toHaveBeenCalled();
        expect(messgeService.moveMessages).not.toHaveBeenCalled();
        expect(applicationService.clearSelectedMessage).not.toHaveBeenCalled();
      });
    });
    test('toggleMessageSeen, messageService setMessagesSeen function invoked', () => {
      // Given
      const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application}};
      initialState.application.selectedMessage = {seen: false};
      applicationService.clearSelectedMessage = jest.fn();
      messgeService.setMessagesSeen = jest.fn();
      const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
      const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

      // When
      topBar.props().toggleMessageSeen();

      // Then
      expect(messgeService.setMessagesSeen).toHaveBeenCalledTimes(1);
      expect(applicationService.clearSelectedMessage).toHaveBeenCalledTimes(1);
    });
    describe('deleteMessages', () => {
      test('Selected messages and selected folder (!== trash), messageService moveMessages function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application},
          folders: {...INITIAL_STATE.folders}, messages: {...INITIAL_STATE.messages}};
        initialState.application.selectedFolderId = '1337';
        initialState.folders.explodedItems = {1337: {}, trash: {type: FolderTypes.TRASH}};
        initialState.messages.cache['1337'] = new Map([{uid: 1}, {uid: 3}].map(m => [m.uid, m]));
        initialState.messages.selected = [1];
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessages();

        // Then
        expect(messgeService.moveMessages).toHaveBeenCalledTimes(1);
        expect(messgeService.deleteMessages).not.toHaveBeenCalled();
      });
      test('Selected message and selected folder (== trash), messageService deleteMessages function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application},
          folders: {...INITIAL_STATE.folders}, messages: {...INITIAL_STATE.messages}};
        initialState.application.selectedFolderId = '1337';
        initialState.folders.explodedItems = {1337: {type: FolderTypes.TRASH}};
        initialState.messages.cache['1337'] = new Map([{uid: 1}, {uid: 3}].map(m => [m.uid, m]));
        initialState.messages.selected = [1];
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessages();

        // Then
        expect(messgeService.deleteMessages).toHaveBeenCalledTimes(1);
        expect(messgeService.moveMessages).not.toHaveBeenCalled();
      });
      test('NO selected message, no function invoked', () => {
        // Given
        const initialState = {...INITIAL_STATE};
        messgeService.moveMessages = jest.fn();
        messgeService.deleteMessages = jest.fn();
        const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
        const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

        // When
        topBar.props().deleteMessages();

        // Then
        expect(messgeService.deleteMessages).not.toHaveBeenCalled();
        expect(messgeService.moveMessages).not.toHaveBeenCalled();
      });
    });
    test('setMessagesSeen,' +
      'with selected messages and selected folder, messageService setMessagesSeen function invoked', () => {
      // Given
      const initialState = {...INITIAL_STATE, application: {...INITIAL_STATE.application},
        folders: {...INITIAL_STATE.folders}, messages: {...INITIAL_STATE.messages}};
      initialState.application.selectedFolderId = '1337';
      initialState.folders.explodedItems = {1337: {type: FolderTypes.TRASH}};
      initialState.messages.cache['1337'] = new Map([{uid: 1}, {uid: 3}].map(m => [m.uid, m]));
      initialState.messages.selected = [1];
      messgeService.setMessagesSeen = jest.fn();
      const props = {sideBarToggle: () => {}, sideBarCollapsed: false};
      const topBar = shallow(<TopBarConnected store={createMockStore(initialState)} {...props}/>);

      // When
      topBar.props().setMessagesSeen();

      // Then
      expect(messgeService.setMessagesSeen).toHaveBeenCalledTimes(1);
    });
  });
});
