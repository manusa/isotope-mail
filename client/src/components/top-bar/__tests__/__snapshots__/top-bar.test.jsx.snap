// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBar component test suite Snapshot render TopBar with new message and no selected message and no selected folder, should render in edit mode without buttons 1`] = `
<header
  className="
      mdc-top-app-bar--with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <TopBarMessageEditor
    collapsed={false}
    sideBarToggle={[MockFunction]}
    title="Isotope Mail Client - TITLE"
  />
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected folder (and messageIds) and no new message and no selected message, should render in list mode with delete and mark read buttons 1`] = `
<header
  className="
      mdc-top-app-bar--with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <Translate(TopBarMessageList)
    collapsed={false}
    onDeleteClick={[Function]}
    onMarkReadClick={[Function]}
    onMarkUnreadClick={[Function]}
    selectedMessages={
      Array [
        Object {
          "uid": 1,
        },
      ]
    }
    sideBarToggle={[MockFunction]}
    title="SelectedFolder - Isotope Mail Client - TITLE"
  />
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected folder and no new message and no selected message, should render in list mode 1`] = `
<header
  className="
      mdc-top-app-bar--with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <Translate(TopBarMessageList)
    collapsed={false}
    onDeleteClick={[Function]}
    onMarkReadClick={[Function]}
    onMarkUnreadClick={[Function]}
    selectedMessages={Array []}
    sideBarToggle={[MockFunction]}
    title="SelectedFolder - Isotope Mail Client - TITLE"
  />
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;

exports[`TopBar component test suite Snapshot render TopBar with selected message and no new message and no selected folder, should render in view mode with buttons 1`] = `
<header
  className="
      mdc-top-app-bar--with-side-bar
      mdc-top-app-bar mdc-top-app-bar--fixed"
>
  <Translate(TopBarMessageViewer)
    collapsed={false}
    onDeleteClick={[Function]}
    outboxEmpty={true}
    sideBarToggle={[MockFunction]}
  />
  <Translate(ConfirmDeleteFromTrashDialog)
    cancelAction={[Function]}
    deleteAction={[Function]}
    visible={false}
  />
</header>
`;
